### PR TITLE
Stress issue

### PR DESCRIPTION
**Pull request by ezrayst**
change `point` class
change `gmsh.tcc` structure a little

### CODE REVIEW

File: include/gmsh.h [hunk 1]
```diff
@@ -24,6 +24,9 @@ class GMSH : public Mesh {
   //! Compute stresses
   void compute_stresses();
 
+  //! Make vector of point class
```
Comment (kks32) [inline include/gmsh.h]: Change to `Generate a vector of material points`

File: include/gmsh.h [hunk 2]
```diff
@@ -24,6 +24,9 @@ class GMSH : public Mesh {
   //! Compute stresses
```
Comment (kks32) [inline include/gmsh.h:25]: Class needs a constructor.

File: include/gmsh.h [hunk 3]
```diff
@@ -33,21 +36,25 @@ class GMSH : public Mesh {
...
 
-  //! Return a vector of stresses
-  std::vector<std::array<double, 6>> stress() const { return stress_; }
-
  private:
   //! Number of vertices
   unsigned nvertices_;
 
-  //! Vector of vertices 
+  //! Vector of id
```
Comment (kks32) [inline include/gmsh.h]: Change to `vector of vertex ids`

File: include/gmsh.h [hunk 4]
```diff
@@ -33,21 +36,25 @@ class GMSH : public Mesh {
...
+  std::vector<unsigned> vertid_;
+
+  //! Vector of coordinates of vertices
+  std::vector<std::array<double, 3>> coords_;
+
+  //! Vector of vertices
```
Comment (kks32) [inline include/gmsh.h]: If I have vertices, why do I need a vector of vertex ids and `coord_`s separately?

File: include/gmsh.tcc [hunk 5]
```diff
@@ -27,25 +24,33 @@ void GMSH::read_vertices(const std::string& filename) {
...
       std::getline(infile, line);
       std::istringstream istream(line);
 
       if (line.find('#') == std::string::npos && line != "") {
         //! Coordinates of vertex
-        std::array<double, ndimension> vertex;
-        
         istream >> vertid;
-        istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
+        istream >> coord.at(0) >> coord.at(1) >> coord.at(2);
```
Comment (kks32) [inline include/gmsh.tcc]: This is the vertex, why is it being replaced by coord?

File: include/gmsh.tcc [hunk 6]
```diff
@@ -126,7 +131,7 @@ void GMSH::read_elements(const std::string& filename) {
   std::cout << "Number of Elements: " << elements_.size() << '\n';
 }
 
-//! Compute stresses
+//! \brief compute initial stress from points
 void GMSH::compute_stresses() {
 
   double density = 22;
```
Comment (kks32) [inline include/gmsh.tcc]: `const`.

File: include/gmsh.tcc [hunk 7]
```diff
@@ -110,14 +115,14 @@ void GMSH::read_elements(const std::string& filename) {
...
         //! \details If element type not == to Tdim, skip element
-        if (elementtype != ndimension) {
-          istream >> line;
-        } else {
-          istream >> elementarray.at(0) >> elementarray.at(1) >>
-              elementarray.at(2) >> elementarray.at(3);
-          elements_.emplace_back(
-              new Point<ndimension>(elementid, elementarray));
-        }
+        // if (elementtype != ndimension) {
```
Comment (kks32) [inline include/gmsh.tcc]: Why is this being commented out?

File: include/io.h [hunk 8]
```diff
@@ -15,30 +15,28 @@ class IO {
...
+    } catch (const std::ifstream::failure& except) {
       std::cerr << "Exception opening/reading mesh file";
     }
     meshfile.close();
-    
+
     //! Material point and stresses
-    vertices_filename_ = "material_points.txt";
-    stress_filename_ = "initial_stresses.txt";
+    vertices_filename_ = "../bin/material_points.txt";
```
Comment (kks32) [inline include/io.h]: No! What happens if I run the code in the root directory? Never do this.

File: include/io.h [hunk 9]
```diff
@@ -15,30 +15,28 @@ class IO {
...
+    stress_filename_ = "../bin/initial_stresses.txt";
   }
 
   //! Write vertices
   void write_vertices(const std::vector<std::shared_ptr<Point<3>>>& vertices);
 
   //! Write stresses
-  void write_stresses(const std::vector<std::array<double, 6>>& stresses);
```
Comment (kks32) [inline include/io.h]: Write stresses should get a vector of stresses, not vertices.

**Review by kks32 — CHANGES_REQUESTED** [review]
@ezrayst You might want to revert most of your changes.

File: include/gmsh.tcc [hunk 10]
```diff
@@ -138,13 +143,25 @@ void GMSH::compute_stresses() {
...
   //! Note that tau (shear stress) is assumed 0
-  for (const auto& point : vertices_) {
-    ver_stress =
-        conv_factor * (-(max_height - point->coordinates().at(2))) * density;
+  for (const auto& point : coords_) {
+    ver_stress = conv_factor * (-(max_height - point.at(2))) * density;
     hor_stress = ver_stress * k0;
     std::array<double, 6> stress{hor_stress, hor_stress, ver_stress, 0, 0, 0};
-    stress_.emplace_back(stress);
+    stress_.push_back(stress);
```
Comment (kks32) [inline include/gmsh.tcc]: Why was this reverted back to `push_back` from `emplace_back`?

File: include/gmsh.tcc [hunk 11]
```diff
@@ -138,13 +143,25 @@ void GMSH::compute_stresses() {
...
+  std::cout << "Computed initial stresses.\n";
+}
+
+//! \brief generate material points
+void GMSH::generate_points() {
+
+  const unsigned ndimension = 3;
+
+  for (unsigned i = 0; i < nvertices_; ++i) {
+    vertices_.emplace_back(
```
Comment (kks32) [inline include/gmsh.tcc]: Is this just so that you can add `stresses` to points?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.tcc [hunk 12]
```diff
@@ -138,13 +143,25 @@ void GMSH::compute_stresses() {
 
   //! Loop through the points to get vertical and horizontal stresses
   //! Note that tau (shear stress) is assumed 0
-  for (const auto& point : vertices_) {
-    ver_stress =
-        conv_factor * (-(max_height - point->coordinates().at(2))) * density;
+  for (const auto& point : coords_) {
```
Comment (kks32) [inline include/gmsh.tcc:181]: Only change required is this, when you find the stress, assign stress to the right material point.

File: include/gmsh.tcc [hunk 13]
```diff
@@ -138,13 +143,25 @@ void GMSH::compute_stresses() {
...
+    stress_.push_back(stress);
+  }
+
+  std::cout << "Computed initial stresses.\n";
+}
+
+//! \brief generate material points
+void GMSH::generate_points() {
+
+  const unsigned ndimension = 3;
```
Comment (kks32) [inline include/gmsh.tcc:194]: Why is the dimension defined here? It should be a global variable, and possibly a template argument.

File: src/main.cc [hunk 14]
```diff
@@ -22,9 +22,12 @@ int main(int argc, char** argv) {
     mesh->read_vertices(io->mesh_file_name());
     mesh->compute_stresses();
 
+    //! Generate material points
+    mesh->generate_points();
+
     //! Write vertices and stresses
     io->write_vertices(mesh->vertices());
-    io->write_stresses(mesh->stress());
+    io->write_stresses(mesh->vertices());
```
Comment (kks32) [inline src/main.cc]: you can't return all the vertices, when all you need is stresses.

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/io.tcc [hunk 15]
```diff
@@ -28,24 +28,26 @@ void IO::write_vertices(
 }
 
 //! \brief Write initial stresses of material points
-//! \param[in[ stresses Initial stress of material points
-void IO::write_stresses(const std::vector<std::array<double, 6>>& stresses) {
+//! \param[in] stresses Initial stress of material points
+void IO::write_stresses(
+    const std::vector<std::shared_ptr<Point<3>>>& vertices) {
```
Comment (kks32) [inline include/io.tcc]: If you are changing function arguments, you need to change the comments! Anyway, this should a vector of stresses.

File: include/point.h [hunk 16]
```diff
@@ -9,19 +9,30 @@ template <unsigned Tdim>
 class Point {
  public:
   //! \brief Constructor with id and coordinates
-  Point(unsigned id, const std::array<double, Tdim>& coord);
+  Point(unsigned id, const std::array<double, Tdim>& coord,
```
Comment (kks32) [inline include/point.h:12]: Not all points have stresses, like a vertex. You can derive a material point class.

File: include/point.h [hunk 17]
```diff
@@ -9,19 +9,30 @@ template <unsigned Tdim>
...
+        const std::array<double, 6>& stress);
 
   //! Returns the id of the vertex
   unsigned id() const { return id_; }
 
   //! Returns coordinates of the vertex
   std::array<double, Tdim> coordinates() const { return coordinates_; }
 
+  //! Returns initial stress of the vertex
+  std::array<double, 6> stress() const { return stress_; }
```
Comment (kks32) [inline include/point.h:22]: You need the point class to hold the vertex info, or is it being done in the Mesh class?

**Review by kks32 — CHANGES_REQUESTED** [review]
Create a vector of `vertices` and a vector of `elements` in the mesh class. If you want to use `Point` to store MPM points.

File: include/gmsh.tcc (hunk 7 above)
Comment (ezrayst) [reply include/gmsh.tcc]: I need to generate the stress to construct `point`. Let me just comment it out first then I will come back to the `elements`

File: include/io.h (hunk 8 above)
Comment (ezrayst) [reply include/io.h]: Then, do I need to add the feature of the directory for the output files?

File: include/gmsh.tcc (hunk 10 above)
Comment (ezrayst) [reply include/gmsh.tcc]: I think the better way is to make `stress` inside `emplace_back`. What do you think?

File: include/gmsh.tcc (hunk 13 above)
Comment (ezrayst) [reply include/gmsh.tcc:194]: The code currently used it this way. It will be changed to a template argument later

File: src/main.cc (hunk 14 above)
Comment (ezrayst) [reply src/main.cc]: In this case `vertices` are `point class` with `stress` stored inside them

File: include/gmsh.tcc (hunk 11 above)
Comment (ezrayst) [reply include/gmsh.tcc]: Yes! I designed the constructor of `points` to contain `id`, `coord`, and `stress`.

File: include/gmsh.tcc (hunk 7 above)
Comment (kks32) [reply include/gmsh.tcc]: Elements have nothing to do with points. Vertices and points are not interchangable.

**Pull request closed without merging**